File: src/main/com/intellij/lang/jsgraphql/schema/builder/GraphQLObjectTypeCompositeDefinition.java
package com.intellij.lang.jsgraphql.schema.builder;

import com.intellij.lang.jsgraphql.types.language.Directive;
import com.intellij.lang.jsgraphql.types.language.FieldDefinition;
import com.intellij.lang.jsgraphql.types.language.ObjectTypeDefinition;
import com.intellij.lang.jsgraphql.types.language.ObjectTypeExtensionDefinition;
import com.intellij.util.containers.ContainerUtil;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static com.intellij.lang.jsgraphql.schema.GraphQLTypeDefinitionUtil.*;

public class GraphQLObjectTypeCompositeDefinition
  extends GraphQLExtendableCompositeDefinition<ObjectTypeDefinition, ObjectTypeExtensionDefinition> {
  @NotNull
  @Override
  protected ObjectTypeDefinition mergeDefinitions(@NotNull List<ObjectTypeDefinition> sourceDefinitions) {
    List<Directive> directives = new ArrayList<>();
    Map<String, FieldDefinition> fieldDefinitions = new LinkedHashMap<>();

    for (ObjectTypeDefinition definition : sourceDefinitions) {
      directives.addAll(definition.getDirectives());
      mergeNodes(fieldDefinitions, mapNamedNodesByKey(definition.getFieldDefinitions()));
    }

    ObjectTypeDefinition definition = ContainerUtil.getFirstItem(sourceDefinitions);
    return definition.transform(builder ->
                                  builder
                                    .directives(directives)
                                    .fieldDefinitions(toList(fieldDefinitions))
                                    .sourceNodes(sourceDefinitions)
    );
  }
}

File: src/main/com/intellij/lang/jsgraphql/schema/builder/GraphQLObjectTypeCompositeDefinition.kt
package com.intellij.lang.jsgraphql.schema.builder

import com.intellij.lang.jsgraphql.types.language.Directive
import com.intellij.lang.jsgraphql.types.language.FieldDefinition
import com.intellij.lang.jsgraphql.types.language.ObjectTypeDefinition
import com.intellij.lang.jsgraphql.types.language.ObjectTypeExtensionDefinition

class GraphQLObjectTypeCompositeDefinition
  : GraphQLExtendableCompositeDefinition<ObjectTypeDefinition, ObjectTypeExtensionDefinition>() {

  override fun mergeDefinitions(sourceDefinitions: List<ObjectTypeDefinition>): ObjectTypeDefinition {
    val directives = mutableListOf<Directive>()
    val fieldDefinitions = mutableMapOf<String, FieldDefinition>()

    for (definition in sourceDefinitions) {
      directives.addAll(definition.directives)
      mergeNodes(fieldDefinitions, mapNamedNodesByKey(definition.fieldDefinitions))
    }

    return sourceDefinitions.first().transform { builder ->
      builder
        .directives(directives)
        .fieldDefinitions(fieldDefinitions.values.toList())
        .sourceNodes(sourceDefinitions)
    }
  }
}
